core/templates/dev/head/components/ck-editor-helpers/ck-editor-5-rte.directive.ts

```diff
@@ -0,0 +1,93 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Directive for CK Editor 5.
+ */
+
+require('services/ContextService.ts');
+require('services/RteHelperService.ts');
+
+const ClassicEditor = require(
+  '@ckeditor/ckeditor5-build-classic/build/ckeditor.js');
+
+var oppia = require('AppInit.ts').module;
+
+oppia.directive('ckEditor5Rte', [
+  'ContextService', 'RteHelperService', 'PAGE_CONTEXT',
+  function(ContextService, RteHelperService, PAGE_CONTEXT) {
+    return {
+      restrict: 'E',
+      scope: {
+        uiConfig: '&'
+      },
+      template: '<div><div></div>' +
+                '<div contenteditable="true" class="oppia-rte">' +
+                '</div></div>',
+      require: '?ngModel',
+
+      link: function(scope: ICustomScope, el, attr, ngModel) {
+        var _RICH_TEXT_COMPONENTS = RteHelperService.getRichTextComponents();
+        var names = [];
+        var icons = [];
+        var canUseFs = ContextService.getPageContext() ===
+          PAGE_CONTEXT.EXPLORATION_EDITOR;
+        _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+          if (!((scope.uiConfig() &&
```

Comment:
This is hard to understand, please make sure your code is readable and easy to follow "at a glance".

I suggest breaking this long clause out into separate variables, e.g. 

`var componentRequiresFsButFsCannotBeUsed = !canUseFs && componentDefn.requiresFs`

(perhaps with a shorter varname that indicates conceptually what you're checking) and use those.

(Though I am a bit confused about the one I just wrote above. If you can't use the filesystem and the component requires the filesystem then why are you including it?)

Reply:
Hi @seanlip @NishealJ `canUseFs` shows whether file system can be used which is only possible when page context is editor mode. Further the component which requires fs is only image (We cannot add image while adding suggestions). So, a component should be disabled only when we cannot use fs and the component requires fs. In all other cases, we should add the component.That's the reason for the check before adding the component. 

Also the condition is a negation check `not (fs cannot be used and component requires fs)`. If the negation is true, component will be added. I think the confusion is due to the long condition in a single line. So, it is a good idea to split up as suggested by @seanlip!

Reply:
Thanks, @bansalnitish also regarding the image plugin how we upload the images we get from the custom image plugin which we are using now? seems like to migrate the image plugin we need to make an upload adapter (CK5 concept refer [here](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/upload-adapter.html)) do you have any other suggestions for this?

Reply:
Hi @NishealJ -- I'm not sure I understand your question to @bansalnitish, but if the question is about how to implement the image plugin in CK5, I think you would need to find a solution (@bansalnitish may be able to offer suggestions, but I don't think he worked with CK5 much, so don't block on them).

For the way the current image plugin works, take a look at extensions/rte_components in Oppia, and look at the code for Image. It's implemented as a standard component to match all the other custom components, and ideally we do the same thing for CK5 -- i.e. have custom components that we maintain that all follow the same "format".

Reply:
Hi @NishealJ, I am not sure if I understand your question clearly. We have added a custom plugin and widget for each rich text component (check `ck-editor-4-widgets.initializer.ts`). The widget on being clicked opens the customization modal for that component and then user can enter required details in the modal. Can you please  explain why do you need extra upload adapter for that?

Reply:
I was reading about it [here](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/upload-adapter.html) I've started the work for plugin migration and I'm finding the APIs which can initialize the rich text components as ck-widgets we had an API for the same in CKEditor 4